### ISSUE TITLE
Avoid multiple parallel network requests for the same URL

### ISSUE THREAD

**Issue by Kazachenko1998** [coil-kt/coil#1461]
I am developing an application where there can be many identical pictures on one screen (SVG from the server).
When I open the application for the first time, I see multiple parallel requests behind the same picture.

I think it's worth adding a pool of downloads, and if this URL is already being processed, do not make this request again, but wait for the first one to complete and use its result.

```
override fun onCreate(savedInstanceState: Bundle?) {
        CODE
        Coil.setImageLoader(
            ImageLoader
                .Builder(this)
                .components { add(SvgDecoder.Factory()) }
                .crossfade(true)
                .build()
        )
        CODE
}

```

```
AsyncImage(
    placeholder = rememberAsyncImagePainter(R.drawable.ic_plchldr),
    error = rememberAsyncImagePainter(R.drawable.ic_plchldr),
    model = URL,
    contentDescription = URL,
    contentScale = ContentScale.Crop,
    modifier = Modifier.fillMaxSize()
)

```
<img width="1464" alt="image" src="https://user-images.githubusercontent.com/22072948/190662362-1714d6be-4fee-4e13-b0d6-48d24a72cc86.png">

**Comment by colinrtwhite** [comment]
I agree this would be nice to have. Ideally, `HttpUriFetcher` should reference count the number of responses waiting for its network request. We probably can't merge requests that only share the same URL - we'll need request headers to match as well (at least by default).

**Comment by colinrtwhite** [comment]
Thinking about this more we should also provide a configurable "keep alive" duration for requests. That way the user can configure how long to wait before cancelling a network request if there's no more consumer for the result.

This would be useful for `RecyclerView`s and other cases where a request can be restarted multiple times in quick succession.

**Comment by gargVader** [comment]
What is the status on this?

**Comment by Airsaid** [comment]
How is the progress on this matter?

**Comment by Airsaid** [comment]
> How is the progress on this matter?

I tried using an interceptor as a temporary workaround for the issue while waiting for an official fix:
```kotlin
/**
 * Merge the same requests to avoid multiple parallel network requests for the same URL.
 *
 * You can set the [mergeFilter] to determine which requests need to be merged,
 * the default is to merge the network requests.
 *
 * You can set the [keyGenerator] to generate the key for the request,
 * the default is to use the request data as the key.
 *
 * Fix the [issue](https://github.com/coil-kt/coil/issues/1461).
 *
 * @param mergeFilter A function that determines which requests need to be merged.
 * @param keyGenerator A function that generates the key for the request.
 * @author airsaid
 */
class MergeRequestInterceptor(
  private val mergeFilter: (ImageRequest) -> Boolean = {
    it.isNetUrlRequest()
  },
  private val keyGenerator: (ImageRequest) -> String = {
    it.generateKey()
  }
) : Interceptor {

  private val pendingRequests = ConcurrentHashMap<String, CompletableDeferred<ImageResult>>()

  override suspend fun intercept(chain: Interceptor.Chain): ImageResult {
    val req = chain.request
    if (!mergeFilter(req)) {
      return chain.proceed(req)
    }

    val requestKey = keyGenerator(req)
    val newDeferred = CompletableDeferred<ImageResult>()
    val existingDeferred = pendingRequests.putIfAbsent(requestKey, newDeferred)

    // If there is a pending request, we should wait for it
    return if (existingDeferred != null) {
      val result = existingDeferred.await()
      result
    } else {
      // Start a new request
      try {
        val result = chain.proceed(chain.request)
        newDeferred.complete(result)
        result
      } catch (e: Exception) {
        newDeferred.completeExceptionally(e)
        throw e
      } finally {
        // Remove the request from the pending map
        pendingRequests.remove(requestKey)
      }
    }
  }
}

private fun ImageRequest.isNetUrlRequest(): Boolean {
  val data = data
  if (data is Uri) {
    return data.isNetworkUri
  }
  if (data is String) {
    val uri = data.toUri()
    return uri.isNetworkUri
  }
  return false
}

private fun ImageRequest.generateKey(): String {
  return data.toString()
}
```

**Comment by Firsto** [comment]
> > How is the progress on this matter?
> 
> I tried to fix the issue through an interceptor:
> 

I tried to use it. But...
Not working anyway.

![image](https://github.com/user-attachments/assets/2e03ef81-d4bb-4d40-b7ba-95cbefa1401c)

**Comment by lukaszkalnik** [comment]
> Additionally - we can only do this for requests that are going to be written to the disk cache. If they're not being written to the disk cache then we can only read the response body once. There's a few other edge cases that means this would have to be opt-in at least for a while.

Why only disk cache? If it's written to the memory cache, isn't it still available to the other requests?

**Comment by colinrtwhite** [comment]
@lukaszkalnik Returning the result of first network request from the memory cache as the result of the second makes a couple assumptions:

- We would have to have a way to have the first request wait for `NetworkFetcher` to finish and then also wait for the relevant `Decoder` to finish and write its result to the memory cache. This adds a lot of coupling between these components, which we should avoid.
- We assume that the result of the first network request will be written to the memory cache (`ImageRequest` params can disable this).
- We assume that the second request will be satisfied by the first request's image in the memory cache. This may not be true if the image is too small.
- We assume that the result image can be shared. This isn't true for images with internal state like animated images.
- We assume that the first network request and second network request will be handled by the same `Decoder` and will produce the same image. This is true by default, but `Decoder` is a custom interface and users can add custom logic so this won't hold in 100% of cases.

Overall, waiting for the first network request to finish, decode its result, and add it to the memory cache would add a lot of coupling and extra complexity.

Instead, it's better to handle network request combining strictly inside `NetworkFetcher`:

If a request reaches `NetworkFetcher` while the same request is in flight, we need to return an `ImageSource` for both requests. Network request response bodies can only be read once so it isn't possible to return that same data twice without copying it in memory (which could OOM). Returning two `ImageSource`s that both read from the disk cache avoids this problem completely.

**Comment by lukaszkalnik** [comment]
Thank you for the thorough explanation.

However, from purely a library user point of view, I understand:
- memory cache as cache which lives as long the application/process is open
- disk cache as cache which survives application/process death.

I didn't expect there will be such significant differences in how they work internally.

**Comment by elyesmansour** [comment]
@colinrtwhite and @luciofm thanks for looking into this 🙏 
When do you think we can expect a new version with the changes in this PR? https://github.com/coil-kt/coil/pull/2995

**Comment by JanMalch** [comment]
@colinrtwhite Is there any update on this issue? I would love to have this feature :)

**Comment by luciofm** [comment]
@JanMalch this got merged to main

**Comment by ApoloApps** [comment]
@luciofm commit?? pr??

**Comment by luciofm** [comment]
@ApoloApps #3326

**Comment by colinrtwhite** [comment]
This is now supported in Coil 3.4.0. Big thanks to @luciofm for implementing this!

It's currently not enabled by default so you need to enable it:

```kotlin
ImageLoader.Builder(context)
    .components {
        // Also supported with KtorNetworkFetcherFactory.
        add(OkHttpNetworkFetcherFactory(concurrentRequestStrategy = { DeDupeConcurrentRequestStrategy() })
    }
    .build()
```
